mozilla-release/browser/components/sessionstore/SessionStore.jsm

```diff
@@ -487,6 +487,66 @@ var SessionStore = {
 // Freeze the SessionStore object. We don't want anyone to modify it.
 Object.freeze(SessionStore);
 
+const cliqz_removeDuplicatedEntries = function(tabs, overwriteTabs) {
+  // CLIQZ-SPECIAL: DB-2447
+  // overwriteTabs is always true if "Show Home Page" option is
+  // not selected on Preferences page.
+  // Otherwise overwriteTabs equals false.
+  //
+  // To handle deduplication problem we need to loop through
+  // every single tab and remove any of its' entries which
+  // should not be restored.
+  // What is left afterwards (if any) will be restored later on.
+  // If overwriteTabs is true then we need to make sure there is
+  // at least one entry left after squashing duplicates.
+  // To detect that we need to find the last entry that not be
+  // restored and then get an index of the tab it belongs to.
+  const homePages = HomePage.get().split("|");
+  const shouldNotRestoreEntry = function(entry) {
+    if (homePages.includes(entry.url)) {
+      return true;
+    }
+
+    return homePages.includes(HomePage.getOriginalDefault()) ?
+      CliqzResources.isCliqzPage(entry.url) :
+      false;
+  };
+
+  let lastTabIndex = tabs.length - 1;
+  if (overwriteTabs) {
+    lastTabIndex = (function() {
+      let i = lastTabIndex;
+      while (i >= 0) {
+        let entries = tabs[i].entries || [];
+        if (entries.length > 0) {
+          let lastEntry = entries[entries.length - 1];
+          if (shouldNotRestoreEntry(lastEntry)) {
+            return i;
+          }
+        }
+
+        i -= 1;
+      }
+      return i;
+    })();
+  }
+
+  // Now we filter out any entry starting from lastTabIndex-1 to 0
+  // which should not be restored;
+  return tabs.filter(function(item, index) {
+    // If "Show Home Page" is not selected
+    if (overwriteTabs && index >= lastTabIndex) {
+      return true;
+    }
+
+    item.entries = item.entries.filter(function(entry) {
```

Comment:
You dont need an entry filter here,
you can just check last entry and decide, if tab should be discarded or included